### PR TITLE
nil pointer dereference on IP parsing

### LINKED ISSUE #291: Crashing when validating external services
# Bug description
Crashing of the Capsule controller when validating external services.

Tenant manifest

```yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  annotations:
    capsule.clastix.io/enable-node-listing: "true"
    capsule.clastix.io/enable-node-update: "true"
    capsule.clastix.io/enable-node-deletion: "true"
  name: oil
spec:
  owner: # required
    name: alice
    kind: User
  nodeSelector:
    pool: cmp
  ingressHostnames:
    allowedRegex: "^.clastix.io$"
  namespaceQuota: 9 
  ingressClasses:
    allowed:
    - cmp
    - haproxy
    allowedRegex: ""
  storageClasses:
    allowed:
    - standard
    allowedRegex: ""
  containerRegistries:
    allowed:
    - docker.io
    - quay.io
    allowedRegex: ""
  externalServiceIPs:
    allowed:
    - 8.8.8.8
```

External service:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-evil-service
  namespace: oil-production
spec:
  selector:
    app: my-evil-dns-server
  ports:
    - name: dns
      protocol: UDP
      port: 53
      targetPort: 9053
  externalIPs:
    - 8.8.4.5
```

# How to reproduce
1. As cluster admin, create a tenant as above
2. As Tenant owner, create a namespace
3. As tenant owner, create an external service as above

# Expected behavior
Capsule denies the creation of the external service 

# Logs

```
2021-06-15T11:15:37.024Z        DEBUG   controller-runtime.webhook.webhooks     received request        {"webhook": "/validating-external-service-ips", "UID": "c61ce7ed-f7ac-451f-bd11-88c2a9b126a9", "kind": "/v1, Kind=Service", "resource": {"group":"","version":"v1","resource":"services"}}
2021/06/15 11:15:37 http: panic serving 10.132.0.10:4976: runtime error: invalid memory address or nil pointer dereference
goroutine 2615 [running]:
net/http.(*conn).serve.func1(0xc000291360)
        /usr/local/go/src/net/http/server.go:1824 +0x153
panic(0x1597640, 0x2285150)
        /usr/local/go/src/runtime/panic.go:971 +0x499
net.networkNumberAndMask(0x0, 0x52b46f, 0xc00052fbc9, 0x7, 0xc00052fc40, 0x10, 0x10)
        /usr/local/go/src/net/ip.go:476 +0x5
net.(*IPNet).Contains(0x0, 0xc00052fc40, 0x10, 0x10, 0x10)
        /usr/local/go/src/net/ip.go:500 +0x2f
github.com/clastix/capsule/pkg/webhook/services.(*handler).handleService(0x22d4300, 0x1932670, 0xc000652880, 0x19439d0, 0xc00003b860, 0xc0000edb00, 0xc000d56d50, 0x24, 0x0, 0x0, ...)
        /workspace/pkg/webhook/services/validating.go:85 +0x985
github.com/clastix/capsule/pkg/webhook/services.(*handler).OnCreate.func1(0x1932670, 0xc000652880, 0xc000d56d50, 0x24, 0x0, 0x0, 0xc000981c60, 0x2, 0xc000981c62, 0x7, ...)
        /workspace/pkg/webhook/services/validating.go:98 +0xe8
github.com/clastix/capsule/pkg/webhook.(*handlerRouter).Handle(0xc0000ed980, 0x1932670, 0xc000652880, 0xc000d56d50, 0x24, 0x0, 0x0, 0xc000981c60, 0x2, 0xc000981c62, ...)
        /workspace/pkg/webhook/router.go:49 +0x435
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).Handle(0xc00064f470, 0x1932670, 0xc000652880, 0xc000d56d50, 0x24, 0x0, 0x0, 0xc000981c60, 0x2, 0xc000981c62, ...)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/webhook/admission/webhook.go:140 +0xd5
sigs.k8s.io/controller-runtime/pkg/webhook/admission.(*Webhook).ServeHTTP(0xc00064f470, 0x7f4bb663ecf8, 0xc000cca910, 0xc000a24700)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/webhook/admission/http.go:100 +0xfa5
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1(0x7f4bb663ecf8, 0xc000cca910, 0xc000a24700)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.7.1/prometheus/promhttp/instrument_server.go:40 +0xab
net/http.HandlerFunc.ServeHTTP(0xc00064f6e0, 0x7f4bb663ecf8, 0xc000cca910, 0xc000a24700)
        /usr/local/go/src/net/http/server.go:2069 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1(0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.7.1/prometheus/promhttp/instrument_server.go:100 +0xdf
net/http.HandlerFunc.ServeHTTP(0xc00064f830, 0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /usr/local/go/src/net/http/server.go:2069 +0x44
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2(0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.7.1/prometheus/promhttp/instrument_server.go:76 +0xb2
net/http.HandlerFunc.ServeHTTP(0xc00064f920, 0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /usr/local/go/src/net/http/server.go:2069 +0x44
net/http.(*ServeMux).ServeHTTP(0xc0000ed000, 0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /usr/local/go/src/net/http/server.go:2448 +0x1ad
net/http.serverHandler.ServeHTTP(0xc000124380, 0x1930070, 0xc0003fa7e0, 0xc000a24700)
        /usr/local/go/src/net/http/server.go:2887 +0xa3
net/http.(*conn).serve(0xc000291360, 0x1932718, 0xc000652640)
        /usr/local/go/src/net/http/server.go:1952 +0x8cd
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3013 +0x39b
W0615 11:15:48.319142       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
```

# Additional context

- Capsule version: 

```
2021-06-15T11:10:18.754Z        INFO    setup   Capsule Version v0.1.0-rc1 8d5391b.dev
2021-06-15T11:10:18.754Z        INFO    setup   Build from: https://github.com/clastix/capsule
2021-06-15T11:10:18.754Z        INFO    setup   Build date: 2021-06-15T08:38:45
2021-06-15T11:10:18.754Z        INFO    setup   Go Version: go1.16.5
2021-06-15T11:10:18.754Z        INFO    setup   Go OS/Arch: linux/amd64
```

- Helm Chart version: (`helm list -n capsule-system`)
- Kubernetes version: (`kubectl version`)

### PR DESCRIPTION
**Pull request by prometherion**
Closes #291.

### CODE REVIEW

**Review by bsctl — COMMENTED** [review]
@prometherion 
we got the same error. Checked both the branches: `hotfix/ip` and `issues/291`. Please, check it.

```
~/capsule$ git branch -a
* hotfix/ip
  issues/291
  master
  remotes/origin/HEAD -> origin/master
  remotes/origin/hotfix/ip
  remotes/origin/issues/231
  remotes/origin/issues/243
  remotes/origin/issues/291
  remotes/origin/master
```

```
2021/06/16 06:30:49 http: panic serving 10.132.0.10:44797: runtime error: invalid memory address or nil pointer dereference
goroutine 921 [running]:
net/http.(*conn).serve.func1(0xc000d55cc0)
        /usr/local/go/src/net/http/server.go:1824 +0x153
panic(0x1597640, 0x2285150)
        /usr/local/go/src/runtime/panic.go:971 +0x499
net.networkNumberAndMask(0x0, 0x52b46f, 0xc000c91039, 0x7, 0xc000c91080, 0x10, 0x10)
        /usr/local/go/src/net/ip.go:476 +0x5
net.(*IPNet).Contains(0x0, 0xc000c91080, 0x10, 0x10, 0x10)
        /usr/local/go/src/net/ip.go:500 +0x2f
github.com/clastix/capsule/pkg/webhook/services.(*handler).handleService(0x22d4300, 0x1932750, 0xc000b81340, 0x1943ab0, 0xc00003bf90, 0xc0002cc380, 0xc00068aa80, 0x24, 0x0, 0x0, ...)
        /workspace/pkg/webhook/services/validating.go:85 +0x985
github.com/clastix/capsule/pkg/webhook/services.(*handler).OnCreate.func1(0x1932750, 0xc000b81340, 0xc00068aa80, 0x24, 0x0, 0x0, 0xc000a87dc0, 0x2, 0xc000a87dc2, 0x7, ...)
        /workspace/pkg/webhook/services/validating.go:98 +0xe8
github.com/clastix/capsule/pkg/webhook.(*handlerRouter).Handle(0xc0002cc200, 0x1932750, 0xc000b81340, 0xc00068aa80, 0x24, 0x0, 0x0, 0xc000a87dc0, 0x2, 0xc000a87dc2, ...
```

**Comment by prometherion** [comment]
Thanks @bsctl, last commit is checking if the CIDR is good or not, avoiding the nil pointer dereference.

**Review by bsctl — COMMENTED** [review]
@prometherion 
now it's blocking the right one

```yaml
apiVersion: v1
kind: Service
metadata:
  name: example
  namespace: sample-investigation
spec:
  selector:
    app: MyApp
  ports:
  - protocol: TCP
    port: 80
    targetPort: 9376
  externalIPs:
  - 8.8.4.5
```

```
admission webhook "validating-external-service-ips.capsule.clastix.io" denied the request:
The selected external IPs for the current Service are violating
the following enforced CIDRs: 8.8.8.8, 8.8.4.5
```

**Comment by bsctl** [comment]
Also got this error

```
2021-06-16T18:26:31.077Z        INFO    controllers.Tenant      Role Binding sync result: updated       {"Request.Name": "sample", "name": "namespace-deleter", "namespace": "sample-production"}
2021-06-16T18:26:31.077Z        INFO    controllers.Tenant      Ensuring Namespace count        {"Request.Name": "sample"}
2021-06-16T18:26:31.078Z        DEBUG   controller-runtime.manager.events       Normal  {"object": {"kind":"Tenant","name":"sample","uid":"83b038a8-a6cd-4cb8-a92e-3153485a1a57","apiVersion":"capsule.clastix.io/v1alpha1","resourceVersion":"397589"}, "reason": "sample-production", "message": "Ensuring Capsule RoleBinding namespace-deleter"}
2021-06-16T18:26:31.084Z        INFO    controllers.Tenant      Tenant reconciling completed    {"Request.Name": "sample"}
W0616 18:26:50.939998       1 warnings.go:70] networking.k8s.io/v1beta1 Ingress is deprecated in v1.19+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
2021-06-16T18:26:52.188Z        ERROR   controller-runtime.manager.controller.endpoints Reconciler error        {"reconciler group": "", "reconciler kind": "Endpoints", "name": "example", "namespace": "sample-investigation", "error": "Endpoints \"example\" not found"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99
2021-06-16T18:26:52.188Z        ERROR   controller-runtime.manager.controller.service   Reconciler error        {"reconciler group": "", "reconciler kind": "Service", "name": "example", "namespace": "sample-investigation", "error": "Service \"example\" not found"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99
2021-06-16T18:26:52.641Z        ERROR   controller-runtime.manager.controller.endpointslice     Reconciler error        {"reconciler group": "discovery.k8s.io", "reconciler kind": "EndpointSlice", "name": "example-qfjsh", "namespace": "sample-investigation", "error": "EndpointSlice.discovery.k8s.io \"example-qfjsh\" not found"}
github.com/go-logr/zapr.(*zapLogger).Error
        /go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:302
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:253
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.8.3/pkg/internal/controller/controller.go:216
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:185
k8s.io/apimachinery/pkg/util/wait.UntilWithContext
        /go/pkg/mod/k8s.io/apimachinery@v0.20.2/pkg/util/wait/wait.go:99
```

**Review by bsctl — COMMENTED** [review]
We should make the recorded event message consistent with other similar events

```
recorder.Eventf(&tnt, corev1.EventTypeWarning, "ForbiddenExternalServiceIP", "Service %s/%s external IPs %s are forbidden for the current Tenant", req.Namespace, req.Name, strings.Join(svc.Spec.ExternalIPs, ","))
	return admission.Errored(http.StatusBadRequest, NewExternalServiceIPForbidden(tnt.Spec.ExternalServiceIPs.Allowed))
}

```

**Comment by prometherion** [comment]
@bsctl I found the issue, finally.

According to the schema, the allowed IPs must match the following regex:

https://github.com/clastix/capsule/blob/46fc65a9888dbdbd84165a59c07b52a0ca21d8cc/api/v1alpha1/tenant_types.go#L23-L24

As you can see, the netmask is optional (`(/([0-9]|[1-2][0-9]|3[0-2]))?`) and Capsule was considering the addresses as a valid CIDR.

The proposed hotfix is going to add the `/32` suffix for the bare IPs, ensuring the `net` validator validates the CIDR required for the function.

**Comment by prometherion** [comment]
JFI I noticed a bug in the external IPs handling, fixed by 9e10beb.

tl;dr; an attacker was able to bypass the Service validation when providing a malicious IP.

Example of the attack:

```yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  name: sample
spec:
  nodeSelector:
    pool: cmp
  owner:
    kind: User
    name: alice
  externalServiceIPs:
    allowed:
    - 8.8.8.8

apiVersion: v1
kind: Service
metadata:
  name: example
  namespace: sample-investigation
spec:
  selector:
    app: MyApp
  ports:
  - protocol: TCP
    port: 80
    targetPort: 9376
  externalIPs:
  - 8.8.8.8
  - 1.1.1.1
```

With the said example, I was able to create a Service exposed both as 8.8.8.8 and 1.1.1.1. since the webhook was returning the allowed response at first match rather than checking all the available Service IPs.

**Comment by prometherion** [comment]
> We should make the recorded event message consistent with other similar events

Fixed with d43bda1

**Comment by bsctl** [comment]
@prometherion we should also consider the case of empty allowed IPs slice to prevent _any_ external IP 

```yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  name: sample
spec:
  nodeSelector:
    pool: cmp
  owner:
    kind: User
    name: alice
  externalServiceIPs:
    allowed: []
```

**Comment by prometherion** [comment]
> @prometherion we should also consider the case of empty allowed IPs slice to prevent _any_ external IP

This was already supported, I just enhanced the error message with commit 661ac68.

```
$: cat /tmp/svc.yaml
apiVersion: v1
kind: Service
metadata:
  name: example
spec:
  selector:
    app: MyApp
  ports:
  - protocol: TCP
    port: 80
    targetPort: 9376
  externalIPs:
  - 8.8.4.5

$: cat /tmp/nosvc.yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  name: sample
spec:
  nodeSelector:
    pool: cmp
  owner:
    kind: User
    name: alice
  externalServiceIPs:
    allowed: []

$: kubectl apply -f /tmp/svc.yaml
Error from server: error when creating "/tmp/svc.yaml": admission webhook "validating-external-service-ips.capsule.clastix.io" denied the request: The current Tenant does not allow the use of externall IPs
```

**Review by bsctl — APPROVED** [review]
@prometherion LGTM Thanks for the hard work!
